torch/nestedtensor/nested/nested.py

```diff
@@ -0,0 +1,133 @@
+import torch
+import numbers
+
+
+def _filter_impl(args, kwargs):
+    if kwargs is None:
+        kwargs = {}
+    impl_args = []
+    for a in args:
+        if isinstance(a, torch.NestedTensor):
+            impl_args.append(a._impl)
+        elif torch.is_tensor(a):
+            impl_args.append(a)
+        elif isinstance(a, list):
+            a_impl, _ = _filter_impl(a, {})
+            impl_args.append(a_impl)
+        elif isinstance(a, tuple):
+            a_impl, _ = _filter_impl(a, {})
+            impl_args.append(tuple(a_impl))
+        else:
+            impl_args.append(a)
+    impl_kwargs = {
+        k: v._impl if isinstance(v, torch.NestedTensor) else v
+        for (k, v) in kwargs.items()
+    }
+    return impl_args, impl_kwargs
+
+
+def _wrap_result(result):
+    if isinstance(result, list):
+        return list(_wrap_result(r) for r in result)
+    if isinstance(result, tuple):
+        return tuple(_wrap_result(r) for r in result)
+    return (
+        torch.NestedTensor(result)
+        if torch.is_tensor(result) and torch.is_nt_impl(result)
+        else result
+    )
+
+
+class NestedTensorMeta(type):
+    def __getattr__(cls, name):
+        if getattr(torch.Tensor, name):
+
+            def _wrapped_fn(*args, **kwargs):
+                impl_args, impl_kwargs = _filter_impl(args, kwargs)
+                result = getattr(impl_args[0], name)(*(impl_args[1:]), **impl_kwargs)
+                return _wrap_result(result)
+
+            return _wrapped_fn
+        return cls.__dict__[name]
+
+
+class NestedTensor(metaclass=NestedTensorMeta):
```

Comment:
I hate to be the naming, but let's be the naming guy. The internal representation here matches those being used by the preexisting faster transformer libraries. I am wondering: do they have other names they refer to this representation as (or do they all call it nested tensor)? In particular, "ragged tensor" is the TF nomenclature for what I believe is the same data structure (although, if I am reading https://github.com/NVIDIA/FasterTransformer/blob/main/docs/encoder_guide.md#standard-bert-and-effective-fastertransformer correctly, none of the row encodings in ragged tensor match the id offset encoding; the trt sequence length offset appears to match `row_splits` ? https://www.tensorflow.org/guide/ragged_tensor#raggedtensor_encoding )

Even if we don't decide to change any of the names here, it will be very helpful for users coming from other settings (e.g., preexisting FasterTransformers users or TF users) to know what NestedTensor is referred to in other contexts. But I also wonder: NestedTensor was originally coined when we were imagining arbitrary amounts of nesting, and it has sense been simplified to the form it is here; maybe nested tensor implies more "nesting" than you should actually expect?

Reply:
Any name you want is the name we will use.